Rules/UseUsingScopeModifierInNewRunspaces.cs

```diff
@@ -0,0 +1,430 @@
+                                        .Left
+                                        .Equals(varExpr),
+                    searchNestedScriptBlocks: true)
+                .Select(a => a as VariableExpressionAst);
+        }
+
+        /// <summary>
+        /// FindNonAssignedNonUsingVarAsts: Retrieve variables that are:
+        /// - not assigned before
+        /// - not prefixed with the 'Using' scope modifier
+        /// - not a PowerShell special variable
+        /// </summary>
+        /// <param name="ast"></param>
+        /// <param name="varsInAssignments"></param>
+        private static List<VariableExpressionAst> FindNonAssignedNonUsingVarAsts(
+            Ast ast, IEnumerable<VariableExpressionAst> varsInAssignments)
+        {
+            // Find all variables that are not locally assigned, and don't have $using: scope modifier
+            return ast.FindAll(
+                    predicate: a => a is VariableExpressionAst varAst &&
+                                    !(varAst.Parent is UsingExpressionAst) &&
+                                    varsInAssignments.All(
+                                        b => b.VariablePath.UserPath != varAst.VariablePath.UserPath) &&
+                                    !Helper
+                                        .Instance
+                                        .HasSpecialVars(varAst.VariablePath.UserPath),
+                    searchNestedScriptBlocks: true)
+                .Select(a => a as VariableExpressionAst).ToList();
+        }
+
+        /// <summary>
+        /// GetSuggestedCorrections: Retrieves a CorrectionExtent for a given variable
+        /// </summary>
+        /// <param name="ast"></param>
+        private static IEnumerable<CorrectionExtent> GetSuggestedCorrections(VariableExpressionAst ast)
+        {
+            var varWithUsing = $"$using:{ast.VariablePath.UserPath}";
+            var description = string.Format(
+                CultureInfo.CurrentCulture,
+                Strings.UseUsingScopeModifierInNewRunspacesCorrectionDescription,
+                ast.Extent.Text,
+                varWithUsing);
+
+            return new[]
+            {
+                new CorrectionExtent(
+                    startLineNumber: ast.Extent.StartLineNumber,
+                    endLineNumber: ast.Extent.EndLineNumber,
+                    startColumnNumber: ast.Extent.StartColumnNumber,
+                    endColumnNumber: ast.Extent.EndColumnNumber,
+                    text: varWithUsing,
+                    file: ast.Extent.File,
+                    description: description
+                )
+            };
+        }
+
+        /// <summary>
+        /// GetSessionName: Retrieves the name of the session (that Invoke-Command is run with).
+        /// </summary>
+        /// <param name="commandAst"></param>
+        private static string GetSessionName(CommandAst commandAst)
+        {
+            if (!(commandAst.CommandElements.First<Ast>(

```

Comment:
An index here would be better for performance

Reply:
Can you help me learn what you mean by giving an example please? I fail to see how to do this rewrite.

Reply:
Sure! I see that really `First` takes a predicate, so we can't simply index it, but something like this:

```csharp
private static bool TryGetSessionNameFromInvokeCommand(CommandAst invokeCommandAst, out string sessionName)
{
    // Sift through Invoke-Command parameters to find the value of the -Session parameter
    // Start at 1 to skip the command name
    for (int i = 1; i < invokeCommandAst.CommandElements.Count; i++)
    {
        // We need a parameter
        if (!(invokeCommandAst.CommandElements[i] is CommandParameterAst parameterAst))
        {
            continue;
        }

        // The parameter must be called "Session"
        if (!parameterAst.ParameterName.Equals("Session", StringComparison.OrdinalIgnoreCase))
        {
            continue;
        }

        // If we have a partial AST, ensure we don't crash
        if (i + 1 >= invokeCommandAst.CommandElements.Count)
        {
            break;
        }

        // Session could be dynamically specified
        // We can only deal with string constants
        if (!(invokeCommandAst.CommandElements[i + 1] is StringConstantExpressionAst stringAst))
        {
            break;
        }

        sessionName = stringAst.Value;
        return true;
    }

    sessionName = null;
    return false;
}
```

Reply:
Thanks for this. Almost there now, tbc

Reply:
Done. I have changed the last test, because `-Session` expects a Session object, usually supplied by a variable. The name of the variable is now the entry we will use to record what belongs to which session.